### PR TITLE
Secrets: Add fallback to plugin on secrets kvstore

### LINKED ISSUE #52564: Implement fallback to sql store while migration to plugin is occurring

### PR DESCRIPTION
**Pull request by guicaulada**
**What this PR does / why we need it**:

This PR adds a fallback SecretsKVStore using the default SQL implementation to the plugin SecretsKVStore.

This is needed to allow for secret retrieval during a background migration to the plugin, the background migration will be implemented to speed up Grafana startup given the plugin have a limited response time speed.

**Which issue(s) this PR fixes**:

Fixes #52564 

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/secrets/kvstore/kvstore.go [hunk 1]
```diff
@@ -51,19 +50,20 @@ func ProvideService(
...
-			store = &secretsKVStorePlugin{
+			pluginStore := NewCachedKVStore(&secretsKVStorePlugin{
 				secretsPlugin:                  secretsPlugin,
 				secretsService:                 secretsService,
 				log:                            logger,
 				kvstore:                        namespacedKVStore,
 				backwardsCompatibilityDisabled: features.IsEnabled(featuremgmt.FlagDisableSecretsCompatibility),
-			}
+			}, 5*time.Second, 5*time.Minute)
+			return NewKVStoreWithFallback(pluginStore, defaultStore), nil
```
Comment (mmandrus) [inline pkg/services/secrets/kvstore/kvstore.go:60]: yo dawg, I heard you like wrappers...

File: pkg/services/secrets/kvstore/fallback.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+type KVStoreWithFallback struct {
+	log      log.Logger
+	store    SecretsKVStore
+	fallback SecretsKVStore
+}
+
+func NewKVStoreWithFallback(store SecretsKVStore, fallback SecretsKVStore) *KVStoreWithFallback {
+	return &KVStoreWithFallback{
+		log:      log.New("secrets.kvstore"),
```
Comment (mmandrus) [inline pkg/services/secrets/kvstore/fallback.go:17]: pass existing logger into here maybe?

File: pkg/services/secrets/kvstore/fallback.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		store:    store,
+		fallback: fallback,
+	}
+}
+
+func (kv *KVStoreWithFallback) Get(ctx context.Context, orgId int64, namespace string, typ string) (string, bool, error) {
+	value, ok, err := kv.store.Get(ctx, orgId, namespace, typ)
```
Comment (mmandrus) [inline pkg/services/secrets/kvstore/fallback.go:24]: 'ok' should be 'exists'

File: pkg/services/secrets/kvstore/fallback.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil || !ok {
+		value, ok, err := kv.fallback.Get(ctx, orgId, namespace, typ)
+		kv.log.Debug("failed to get secret with plugin, using fallback", "orgId", orgId, "type", typ, "namespace", namespace, "error", err)
+		return value, ok, err
+	}
+	return value, ok, err
+}
+
+func (kv *KVStoreWithFallback) Set(ctx context.Context, orgId int64, namespace string, typ string, value string) error {
```
Comment (mmandrus) [inline pkg/services/secrets/kvstore/fallback.go:33]: If this is meant to be a fallback when migration is occurring, I'm not sure it makes sense to actually implement set, del, and rename. For example, if you set a key, it won't be in the migrator's list of keys, so it won't get deleted and will stay forever. Therefore, perhaps all of the setters should just call the primary store.

File: pkg/services/secrets/kvstore/fallback.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	return &KVStoreWithFallback{
+		log:      log.New("secrets.kvstore"),
+		store:    store,
+		fallback: fallback,
+	}
+}
+
+func (kv *KVStoreWithFallback) Get(ctx context.Context, orgId int64, namespace string, typ string) (string, bool, error) {
+	value, ok, err := kv.store.Get(ctx, orgId, namespace, typ)
+	if err != nil || !ok {
```
Comment (mmandrus) [inline pkg/services/secrets/kvstore/fallback.go:25]: we might want to guard this a bit more with a check for whether a migration is actually occurring. If we already migrated and then there is an error or exists = false, then we should probably just return the error. Maybe the migrator can set an in-memory flag after it has completed and we can check if it has been set here?

**Review by mmandrus — COMMENTED** [review]
A few conceptual questions here. If this is just intended to be a fallback during migration, I think we need to update the logic a little. If there are other reasons why we would have a fallback, then it's a different story

File: pkg/services/secrets/kvstore/fallback.go (hunk 4 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go:33]: I also considered that.

File: pkg/services/secrets/kvstore/fallback.go (hunk 5 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go:25]: In this case the fallback secret store could be used only for migration, so it shouldn't be returned on kvstore.go

File: pkg/services/secrets/kvstore/fallback.go (hunk 2 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go:17]: Hmm all the other wrappers call log.New() not sure if we should change that now or leave that for a future refactor.

Comment (mmandrus) [reply pkg/services/secrets/kvstore/fallback.go:17]: Okay, yeah it's good as-is then 👍

File: pkg/services/secrets/kvstore/fallback.go (hunk 3 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go:24]: Done!

File: pkg/services/secrets/kvstore/fallback.go (hunk 4 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go:33]: Done!

File: pkg/services/secrets/kvstore/fallback.go (hunk 5 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go]: Added a flag to the struct that migration can use to disable fallback once it's done.

File: pkg/services/secrets/kvstore/fallback.go (hunk 2 above)
Comment (guicaulada) [reply pkg/services/secrets/kvstore/fallback.go]: Left like this for now to follow the same pattern, this could be changed on a future refactor.

**Comment by mmandrus** [comment]
Nice changes. I think it would be within the scope of this task to update the migrator too to update the flag, even if it's not a background service yet. I think it should be good to go after that. May also be worth writing a unit test to make sure that flag gets updated and read properly.

File: pkg/services/secrets/kvstore/fallback.go (hunk 4 above)
Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/fallback.go:34]: I feel like `del` and `rename` would have different results if executed before/during/after migration. 
the problem with having the migration as a background service is that we can to be sure (for now) it is finished. 
How would `del` and `rename` will behave if the secrets are not yet migrated?

File: pkg/services/secrets/kvstore/kvstore.go (hunk 1 above)
Comment (leandro-deveikis) [inline pkg/services/secrets/kvstore/kvstore.go:60]: Every time we have a Secret plugin configured we would need a Fallback? What if migration will not run?

Comment (mmandrus) [reply pkg/services/secrets/kvstore/kvstore.go:60]: Good point, maybe the `useFallback` flag in the wrapper could be set based on the migration flag

Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go:60]: In that case the fallback could be disabled by default, but I think it's easier to return the wrapper than have a bunch of if and elses and return a different version of secret store each time.

In the future once we have LockAndRelease and migration as a background service Migration code should run every time, and the check for if is needed is done inside the Migrate code, which could also set or unset the feedback usage depending on those coditions.

**Comment by guicaulada** [comment]
Going over some of the points mentioned, finding a bit of difficulty to implement this only to migration.

**Comment by guicaulada** [comment]
Rethinking this approach.

**Comment by guicaulada** [comment]
Closed this for now.

**Pull request closed without merging**